lambdas/es/indexer/index.py

```diff
@@ -94,189 +240,278 @@ def extract_text(notebook_str):
                     except (KeyError, json.JSONDecodeError):
-                        print('decoding helium metadata failed')
+                        print("Unable to parse Quilt 'helium' metadata", meta)
 
-                    post_to_es(event_type, size, text, key, meta, version_id)
-                except Exception as e:
-                    # do our best to process each result
-                    print("Exception encountered for record")
-                    print(e)
+                    batch_processor.append(
+                        event_name,
+                        bucket=bucket,
+                        key=key,
+                        ext=ext,
+                        meta=meta,
+                        etag=etag,
+                        version_id=version_id,
+                        last_modified=last_modified,
+                        size=size,
+                        text=text
+                    )
+                except Exception as exc:# pylint: disable=broad-except
+                    print("Fatal exception for record", event_, exc)
                     import traceback
-                    traceback.print_tb(e.__traceback__)
-                    print(msg)
+                    traceback.print_tb(exc.__traceback__)
+            # flush the queue
+            batch_processor.send_all()
 
-    except Exception as e:
-        # do our best to process each result
-        print("Exception encountered for whole Event")
-        print(e)
+    except Exception as exc:# pylint: disable=broad-except
+        print("Fatal exception for message", message, event_, exc)
         import traceback
-        traceback.print_tb(e.__traceback__)
-        print(event)
-        # Fail the lambda so the message is not dequeued.
-        raise e
+        traceback.print_tb(exc.__traceback__)
+        # Fail the lambda so the message is not dequeued
+        raise exc
+
+def retry_s3(
+        operation,
+        context,
+        bucket,
+        key,
+        size,
+        limit=DOC_SIZE_LIMIT_BYTES,
+        *,
+        etag,
+        version_id
+):
+    """retry head or get operation to S3 with; stop before we run out of time.
+    retry is necessary since, due to eventual consistency, we may not
+    always get the required version of the object.
+    """
+    if operation not in ["get", "head"]:
+        raise ValueError(f"unexpected operation: {operation}")
+    if operation == "head":
+        function_ = S3_CLIENT.head_object
+    else:
+        function_ = S3_CLIENT.get_object
+
+    time_remaining = get_time_remaining(context)
+    # use a local function so that we can parameterize to time_remaining
+    @retry(
+        # debug
+        stop=(stop_after_delay(time_remaining) | stop_after_attempt(MAX_RETRY)),
+        wait=wait_exponential(multiplier=2, min=4, max=30)
+    )
+    def call():
+        # we can't use Range= if size == 0 because byte 0 doesn't exist
+        # and we'll get an exception
+        if size == 0:
```

Comment:
I think a dict of params would be better - then you can handle `size==0` and `version_id` independently, instead of having 4 different cases.